cmd/ocsp-responder/main.go

```diff
@@ -32,27 +33,101 @@ import (
 	"github.com/letsencrypt/boulder/sa"
 )
 
-/*
-DBSource maps a given Database schema to a CA Key Hash, so we can pick
-from among them when presented with OCSP requests for different certs.
+// ocspFilter stores information needed to filter OCSP requests (to ensure we
+// aren't trying to serve OCSP for certs which aren't ours), and surfaces one
+// boolean method to determine if a given request should be filtered or not.
+type ocspFilter struct {
+	issuerKeyHashAlgorithm crypto.Hash
+	issuerKeyHashes        [][]byte
+	serialPrefixes         []string
+}
+
+// check returns a descriptive error if the request does not satisfy any of
+// the requirements of an OCSP request, or nil if the request should be handled.
+func (f *ocspFilter) check(req *ocsp.Request) error {
+	if req.HashAlgorithm != f.issuerKeyHashAlgorithm {
+		return fmt.Errorf("Request ca key hash using unsupported algorithm %s: %w", req.HashAlgorithm, bocsp.ErrNotFound)
+	}
+	// Check that this request is for the proper CA
+	match := false
+	for _, keyHash := range f.issuerKeyHashes {
+		if match = bytes.Equal(req.IssuerKeyHash, keyHash); match {
+			break
+		}
+	}
+	if !match {
+		return fmt.Errorf("Request intended for wrong issuer cert %s: %w", hex.EncodeToString(req.IssuerKeyHash), bocsp.ErrNotFound)
+	}
 
-We assume that OCSP responses are stored in a very simple database table,
-with two columns: serialNumber and response
+	serialString := core.SerialToString(req.SerialNumber)
+	if len(f.serialPrefixes) > 0 {
+		match := false
+		for _, prefix := range f.serialPrefixes {
+			if match = strings.HasPrefix(serialString, prefix); match {
+				break
+			}
+		}
+		if !match {
+			return fmt.Errorf("Request serial has wrong prefix: %w", bocsp.ErrNotFound)
+		}
+	}
 
-  CREATE TABLE ocsp_responses (serialNumber TEXT, response BLOB);
+	return nil
+}
 
-The serialNumber field may have any type to which Go will match a string,
-so you can be more efficient than TEXT if you like.  We use it to store the
-serial number in base64.  You probably want to have an index on the
-serialNumber field, since we will always query on it.
+func newFilter(issuerCerts []string, serialPrefixes []string) (*ocspFilter, error) {
+	if len(issuerCerts) < 1 {
+		return nil, errors.New("Filter must include at least 1 issuer cert")
+	}
+	var issuerKeyHashes [][]byte
+	for _, issuerCert := range issuerCerts {
+		// Load the certificate from the file path.
+		caCertDER, err := cmd.LoadCert(issuerCert)
```

Comment:
If we use the `LoadCert` variant in `core`, we don't have to separately call `ParseCertificate` (we should probably remove the LoadCert in cmd).

Reply:
Good call. I've filed #5163 to remove the only other remaining use of `cmd.LoadCert`.